---
title: "Acerv_PopStructure"
author: "Sheila Kitchen"
date: "February 13, 2022"
output:
  pdf_document:
    latex_engine: xelatex
  html_document: default
---

```{r setup, include=FALSE}
.libPaths("/scratch/group/kitchen-group/Rlibs")
knitr::opts_chunk$set(echo = TRUE)
```

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
#Required R-packages for multi-locus genotype calling
library(vcfR)
library(poppr)
library(adegenet)
library(ape)
library(hierfstat)
library("StAMPP")
library(ggplot2)
library(data.table)
library(reshape2)
library(dplyr)
library(tidyverse)
library(boa)
library("sf")
library("rnaturalearth")
library("rnaturalearthdata")
library(scatterpie)
library(mapplots)
library(psych)
library(multcompView)
library(vegan)
library(ggpmisc)
library(geosphere)
library(SNPRelate)
theme_set(theme_bw())
```

#Import the VCF file:
In order to analyze the data, we need to indicate where the `.vcf` file that we generated previously is stored. The `.vcf` file, short for **variant call format**, contains information on differences in the genetic information of particular animals from some reference genome sequence (an outgroup species, in this case). This file format is much more efficient than storing all the genetic information for a given sample (animal, in our case), because much of this is redundant from one individual to another (many of the ~100 million base pairs are the same for many members of a species, besides for the single nucleotide polymorphisms (SNPs) and potential insertions or deletions). Using the `.vcf` to keep track of differences from a reference genome, you can maintain all the relevant (disparate) information on each of a set of samples along with the reference genome, which houses all the redundant information that applies to all samples (its a huge storage space saver!). Full details on the `.vcf` format can be found at the [1000 Genomes Project](http://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/VCF%20%28Variant%20Call%20Format%29%20version%204.0/encoding-structural-variants) website).

Based on the unique multi-locus genotype (MLG) report, I extracted the 57 MLGs from the 72 Acropora cervicornis samples using VCFTOOLS (outside of R). I combined those with the unique A. cervicornis MLGs from another paper [Kitchen et al. 2020](https://rdcu.be/dm7j2) that included additional geographic locations from across the Caribbean (n=121 MLGs) and filtered for a minor allele frequency of 5%.
```{r import}
##################
### Import VCF ###
##################

#Read in VCF file with array SNVs
ACvcf <- read.vcfR("AC_mergeDB_maf0.05.recode.vcf")
ACvcf
```
There is a lot of useful information here on what kind of data we are working with. You can see that there are 6201 variants. We can also see that there are 921 `CHROMs` (chromosomes) that these variants fall on; in the case of a very well annotated and assembled genome like *Drosophila* or humans, these will truly correspond to chromosomes, though for this coral analysis, these represent contigs instead. A **contig** is a contiguous section of DNA given from the assembly process of a genome; they may represent a full chromosome, depending on what data went into the assembly, or some fragmented region of a chromosome. Which chromosome variants fall on is not really the scope of this analysis, since we have yet to achieve a chromosome level assembly, but might be at some point.

Let's now add more information about the samples in this data set.
```{r}
#add population or species information to the genind pop slot
poptab<-read.table("acerv_pop3.txt",
                   check.names=FALSE, header=F, na.strings = c("", "NA"),
                   stringsAsFactors = FALSE, sep="\t",encoding="UTF-8")
colnames(poptab)<-c("row_id","affy_id","user_specimen_id","region", "lat", "long","subregion", "region2","subregion2","assigned_pop")

gindAC <- vcfR2genlight(ACvcf)
gindAC@pop <- as.factor(poptab$region)
strata(gindAC) <- data.frame(poptab[,c(4,7)])
gindAC

```

The **multilocus genotypes** (MLGs) are critically important for tracking dispersal and population structure of organisms, especially those that reproduce clonally (plants, sponges, cnidarians, flatworms, annelids, sea stars, and many more). For such clonal organisms, we expect that offspring of a given animal will have the same genomic variants as their parent (there could be very slight differences due to mutations in the germ line). The data was pre-processed to only retain unique individuals for all the downstream analysis. We covered how to identify MLGs in Case Study 2.

# Population Structure Analyses

For illustration purposes, we will look at a population structure [study](https://doi.org/10.7717/peerj.13854) of reef-building corals called *Acropora cervicornis* and *Acropora palmata* (Fig 1 B,C). For this study, though, a third species, called *Acropora digitifera* was used as an outgroup, and all variant calls were made based on the genome of this third species. The goal of this study was to analyze the population structure of these two coral species in Colombia, a location that has not been incorporated in prior studies. The data was combined with individuals from around the Caribbean: Belize, Curacao, Florida, and the US Virgin Islands.

<img src="https://www.g3journal.org/content/ggg/9/5/1633/F1.large.jpg?width=800&height=600&carousel=1"
    alt="Drawing"
    style="width: 500px;"/>

Figure 1. Coral species used in this genomic variant analysis.</center></font>

For today's activity we will focus on only _A. cervicornis_.

## K-clustering- how many clusters/populations?
Identification of the best K is performed by find.clusters in the adegenet package.This function first transforms the data using PCA. Then, it runs k-means algorithm with increasing values of k and computes associated summary statistics (by default, BIC). This approach looks for a minimal cluster number that best explains the pattern in the data. Adding more clusters than the minimal number that best explain the data gets penalized under this approach because it adds additional parameters to the model. We will take the cluster number that minimized the BIC, though this package has other options for how to do this. They actually recommend that you do it by eye from a plot of the BIC vs cluster number, but we use the automated option here.

```{r ks}

# find optimal k, repeat 10 times for each value
maxK <- 10
myMat <- matrix(nrow=10, ncol=maxK)
colnames(myMat) <- 1:ncol(myMat)
for(i in 1:nrow(myMat)){
  grp <- find.clusters(gindAC, n.pca = 120, choose.n.clust = FALSE,  max.n.clust = maxK)
  myMat[i,] <- grp$Kstat
}

my_df <- melt(myMat)
colnames(my_df)[1:3] <- c("Group", "K", "BIC")
my_df$K <- as.factor(my_df$K)
head(my_df)
```

Now plot the results:
```{r plotK}
p1 <- ggplot(my_df, aes(x = K, y = BIC))
p1 <- p1 + geom_boxplot()
p1 <- p1 + theme_bw()
p1 <- p1 + xlab("Number of groups (K)")
p1

```
There is a decrease in BIC at k= 4, after which the BIC increases. This suggests 4 clusters should be retained.

## PCA
Summarize the variation between the samples using all the SNP data.
```{r PCA}
ac.pca <- glPca(gindAC, nf = 10)
```

Plot of the % variation explained by eigenvectors
```{r}
#plot of the % variation explained by eigenvectors
barplot(100*ac.pca$eig/sum(ac.pca$eig), col = heat.colors(50), main="PCA Eigenvalues")
title(ylab="Percent of variance\nexplained", line = 2)
title(xlab="Eigenvalues", line = 1)
```

Create new data frame with the PC scores per sample:
```{r df}
ac.pca.scores <- as.data.frame(ac.pca$scores)
ac.pca.scores$pop <- pop(gindAC)
head(ac.pca.scores)
```

Extract the variance explained by each principal component retained.
```{r var}
# variance explained by each axis
var_frac <- ac.pca$eig/sum(ac.pca$eig)
var_frac[1:4]
```

Create a vector subpopulations within Colombia for plotting purposes:
```{r}
# Colombia regions/subpopulations
Colombia<-c("Urabá Gulf","Cartagena","Rosario", "San Andrés", "San Bernardo", "Tayrona", "I. Arena")
```

Now we can plot the PCA for the regions within and outside of Colombia.
```{r plotPCA}
#plot the PCA
set.seed(9)
cols2 <- c(`San Andrés`="#003049",`Urabá Gulf`="#284CB8",`San Bernardo`="#7A94E1",`Rosario`="#3370A3",`I. Arena`="#69BAC7",`Tayrona`="#97CFD8",`Florida`="#959177",`Belize`= "#FCBF49",`Cuba`="#9EDFA8", `Curacao`="#F77F00",`Puerto Rico`="#D62828" ,`USVI`="#961B33")

p <- ggplot(ac.pca.scores, aes(x=PC1, y=PC2, color=pop))
p <- p + geom_point(size=2)
p <- p + stat_ellipse(level = 0.95, size = 1)
p <- p + scale_color_manual(values=cols2)
p <- p + theme_bw() +ylab("PC2 (6.1%)") + xlab("PC1 (11.4%)")

p
```

Follow up by only looking at the Colombia samples, the focus of this study.
```{r}
# just Colombia
sub_gindAC<-popsub(gindAC,
  sublist = c("San Andrés","Urabá Gulf","San Bernardo","Rosario","I. Arena","Tayrona")
)

ac.pca <- glPca(sub_gindAC, nf = 10)
ac.pca.scores <- as.data.frame(ac.pca$scores)
ac.pca.scores$pop <- pop(sub_gindAC)

# variance explained by each axis
var_frac <- ac.pca$eig/sum(ac.pca$eig)

set.seed(9)
cols2 <- c(`San Andrés`="#003049",`Urabá Gulf`="#284CB8",`San Bernardo`="#69BAC7",`Rosario`="#961B33" ,`Cartagena`="#D62828",`I. Arena`="#F77F00",`Tayrona`="#FCBF49")
p <- ggplot(ac.pca.scores[ac.pca.scores$pop %in% Colombia,], aes(x=PC1, y=PC2, color=pop))
p <- p + geom_point(size=2)
p <- p + stat_ellipse(level = 0.95, size = 1)
p <- p + scale_color_manual(values=cols2)
p <- p + theme_bw() +ylab("PC2 (2.5%)") + xlab("PC1 (14.5%)")

p

```
Urabá Gulf is a transitional zone between San Andres and the other locations. The two samples from Curacao overlap the transitional zone of the Colombia samples.

## DAPC
We can use a discriminant analysis of principal components to cluster genetically related individuals. This multivariate statistical approach partitions the sample into a between-group and within- group component, in an effort to maximize discrimination between groups. Data is first transformed using a principal components analysis (PCA) and subsequently clusters are identified using discriminant analysis (DA). Unlike STRUCTURE it does not make assumptions on population genetic parameters such as Hardy Weinberg equilibrium. More information can be found [here](http://adegenet.r-forge.r-project.org/files/tutorial-dapc.pdf).

### Using the optimal K determined above
```{r dapc}
#use best k from above
set.seed(9)
ac.dapc <- dapc(gindAC, grp$grp, n.pca=40, n.da=3)
summary(ac.dapc)
```

```{r}
cols<-c("#D62828","#003049","#FCBF49","#EAE2B7")

my_df <- as.data.frame(ac.dapc$ind.coord)
my_df$Group <- ac.dapc$grp
my_df$pop <-poptab$region
head(my_df)

p2 <- ggplot(my_df, aes(x = LD1, y = LD2, color = factor(Group), shape=factor(pop)))
p2 <- p2 + geom_point(size = 3, alpha=0.8)
p2 <- p2 + theme_bw()
p2 <- p2 + scale_color_manual(values=cols)+
  scale_shape_manual(values=c(17,18,19,7,21,22,23,24,25,8,15,9))
p2
```

The first plot we will make is a PCA to split the *A. cervicornis* samples based on region. We will plot the first two principle components, which will show the two orthogonal vectors in the data that explain the most variance.
```{r}
#separate by populations
dapc.results <- as.data.frame(ac.dapc$posterior)
dapc.results$pop <- pop(gindAC)
dapc.results$indNames <- rownames(dapc.results)

#reorganize dataframe for ggplot
dapc.results <- melt(dapc.results)
colnames(dapc.results) <-c("Original_Pop","Sample","Assigned_Pop","Posterior_membership_probability")
dapc.results$userID <- poptab$user_specimen_id

dapc.results$Original_Pop = factor(dapc.results$Original_Pop, levels=c( "Florida","Cuba","Belize","San Andrés","Urabá Gulf","San Bernardo","Rosario","I. Arena","Tayrona","Curacao","Puerto Rico","USVI"))

p <- ggplot(dapc.results, aes(x=userID, y=Posterior_membership_probability, fill=Assigned_Pop))
p <- p + geom_bar(stat='identity',color="darkgrey",size=0.0001)
p <- p + facet_grid(~ Original_Pop, scales = "free_x",space = "free_x")
p <- p + theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5), panel.spacing = unit(0, "lines"))+
    scale_y_continuous(expand = c(0, 0))
p <- p + scale_fill_manual(values=cols)
p

```
We can see the two samples from Urabá Gulf were assigned to pop2 and the rest to pop4.

### Map the assigned populations to their geographic locations
Next, we can map the cluster assignments to the location the samples were collected.
```{r maps}
world <- ne_countries(scale = "medium", returnclass = "sf")
class(world)

poptab$assign1<-ac.dapc$assign==1
poptab$assign1[poptab$assign1=="TRUE"]<-1
poptab$assign1[poptab$assign1=="FALSE"]<-0
poptab$assign2<-ac.dapc$assign==2
poptab$assign2[poptab$assign2=="TRUE"]<-1
poptab$assign2[poptab$assign2=="FALSE"]<-0
poptab$assign3<-ac.dapc$assign==3
poptab$assign3[poptab$assign3=="TRUE"]<-1
poptab$assign3[poptab$assign3=="FALSE"]<-0
poptab$assign4<-ac.dapc$assign==4
poptab$assign4[poptab$assign4=="TRUE"]<-1
poptab$assign4[poptab$assign4=="FALSE"]<-0

ag<-aggregate(cbind(assign1,assign2,assign3,assign4) ~ subregion, data = poptab, sum)

ag2 <-ag %>%
  right_join(poptab %>%
      select("subregion","lat", "long","region"),by='subregion') %>%
  group_by(region) %>%
  distinct(subregion,.keep_all = TRUE) %>%
  mutate(radius=sum(assign1,assign2,assign3,assign4)/15) %>%
  mutate(pop1=sum(assign1),pop2=sum(assign2),pop3=sum(assign3),pop4=sum(assign4), n=sum(assign1,assign2,assign3,assign4)) %>%
  distinct(region,.keep_all = TRUE) %>%
  select(-assign1, -assign2,-assign3,-assign4)

p5<-ggplot(data = world) +
    geom_sf() +
    coord_sf(xlim = c(-60, -90), ylim = c(7, 30), expand = FALSE)

p6 <-p5+ geom_scatterpie(data = ag2,aes(x =long, y=lat, group=region, r=radius), cols=c("pop1","pop2","pop3","pop4"),alpha=0.5)+geom_scatterpie_legend(ag2$radius,x=-68, y=25, n=4)
p6 <- p6 + scale_fill_manual(values=cols)
p6

```
#### Zoom in on the split assignment by subregions in Urabá Gulf
The collection site Bajo Naui has two samples assigned to pop2 and one sample assigned to pop 4.
```{r map-zoom}
p8<-ggplot(data = world) +
    geom_sf() +
    coord_sf(xlim = c(-77.6, -77.1), ylim = c(8.5, 9), expand = TRUE)

ag3 <-ag %>%
  right_join(poptab %>%
      select("subregion","lat", "long","region"),by='subregion') %>%
  group_by(subregion) %>%
  distinct(subregion,.keep_all = TRUE) %>%
  mutate(radius=sum(assign1,assign2,assign3,assign4)/15) %>%
  mutate(pop1=sum(assign1),pop2=sum(assign2),pop3=sum(assign3),pop4=sum(assign4), n=sum(assign1,assign2,assign3,assign4)) %>%
  select(-assign1, -assign2,-assign3,-assign4)

p9<-p8+ geom_scatterpie(data = ag3,aes(x =long, y=lat, group=subregion, r=radius/10), cols=c("pop1","pop2","pop3","pop4"),alpha=.8)
p9 <- p9 + scale_fill_manual(values=cols)
p9

```
Now we can see where the individuals from Urabá Gulf were collected and their population assignment.

### Testing different cluster sizes
We can look for further sub-structure in the data by collection site, although the best K-means cluster is 4.
```{r dapc_multiKs}
my_k <- 2:6

grp_l <- vector(mode = "list", length = length(my_k))
dapc_l <- vector(mode = "list", length = length(my_k))

for(i in 1:length(dapc_l)){
  set.seed(7)
  grp_l[[i]] <- find.clusters(gindAC, n.pca = 40, n.clust = my_k[i])
  dapc_l[[i]] <- dapc(gindAC, pop = grp_l[[i]]$grp, n.pca = 40, n.da = my_k[i])
}

my_df <- as.data.frame(dapc_l[[ length(dapc_l) ]]$ind.coord)
my_df$Group <- dapc_l[[ length(dapc_l) ]]$grp
my_df$pop <-poptab$region
head(my_df)

cols2 <- c(`San Andrés`="#003049",`Urabá Gulf`="#284CB8",`San Bernardo`="#7A94E1",`Rosario`="#3370A3",`I. Arena`="#69BAC7",`Tayrona`="#97CFD8",`Florida`="#959177",`Belize`= "#FCBF49",`Cuba`="#9EDFA8", `Curacao`="#F77F00",`Puerto Rico`="#D62828" ,`USVI`="#961B33")

p2 <- ggplot(my_df, aes(x = LD1, y = LD2, color = factor(pop), shape=factor(Group)))
p2 <- p2 + geom_point(size = 3, alpha=0.8)
p2 <- p2 + theme_bw()
p2 <- p2 + scale_color_manual(values=cols2)+
  scale_shape_manual(values=c(17,18,19,20,21,22,23,24))
p2
```

```{r}
tmp <- as.data.frame(dapc_l[[1]]$posterior)
tmp$K <- my_k[1]
tmp$Genet <- rownames(tmp)
tmp <- melt(tmp, id = c("Genet", "K"))
names(tmp)[3:4] <- c("Group", "Posterior")
tmp$Region <- poptab$region
my_df <- tmp

for(i in 2:length(dapc_l)){
  tmp <- as.data.frame(dapc_l[[i]]$posterior)
  tmp$K <- my_k[i]
  tmp$Genet <- rownames(tmp)
  tmp <- melt(tmp, id = c("Genet", "K"))
  names(tmp)[3:4] <- c("Group", "Posterior")
  tmp$Region <- poptab$region

  my_df <- rbind(my_df, tmp)
}

grp.labs <- paste("K =", my_k)
names(grp.labs) <- my_k
my_df$userID <- poptab$user_specimen_id

my_df$Region = factor(my_df$Region, levels=c("Florida","Cuba","Belize","San Andrés","Urabá Gulf","San Bernardo","Rosario","I. Arena","Tayrona","Curacao","Puerto Rico","USVI"))

cols3 <- c(`1`="#FCBF49",`2`="#003049",`3`="#D62828",`4`="#69BAC7",`5`="#EAE2B7",`6`="#F77F00")
p3 <- ggplot(my_df, aes(x = userID, y = Posterior, fill = Group))
p3 <- p3 + geom_bar(stat = "identity",color="darkgrey",size=0.0001)
p3 <- p3 + facet_grid(K ~ Region, scales = "free_x", space = "free",
                      labeller = labeller(K = grp.labs))
p3 <- p3 + theme_bw()
p3 <- p3 + ylab("Posterior probability of group membership")
p3 <- p3 + theme(legend.position='none')
p3 <- p3 + scale_fill_manual(values=c(cols3))
p3 <- p3 + theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5), panel.spacing = unit(0, "lines"))+
    scale_y_continuous(expand = c(0, 0))
p3

```
At k=5 Cuba, San Andres and Curacao become a separate population from the eastern pops USVI and Puerto Rico.

## ADMIXTURE
ADMIXTURE is a software tool for maximum likelihood estimation of individual ancestries from multilocus SNP genotype datasets. It uses the same statistical model as STRUCTURE but calculates estimates much more rapidly using a fast numerical optimization algorithm. This is key when you have many millions of SNPs (either tool is fine with the number of SNPs we are using today).

Specifically, ADMIXTURE uses a block relaxation approach to alternately update allele frequency and ancestry fraction parameters. Each block update is handled by solving a large number of independent convex optimization problems, which are tackled using a fast sequential quadratic programming algorithm. Convergence of the algorithm is accelerated using a novel quasi-Newton acceleration method. The algorithm outperforms EM algorithms and MCMC sampling methods by a wide margin. 

Every K value was estimated 10 times with a cross-validation procedure. More precisely, the cross-validation procedure partitions all the observed genotypes into v = 10 roughly equally-sized folds. The procedure masks (i.e. converts to
“MISSING”) all genotypes, for each fold in turn. For each fold, the resulting masked dataset G˜ is used to calculate estimates ˜θ = (Q, ˜ P˜). The prediction error is estimated by averaging the squares of the deviance residuals for the binomial model.

Here we can look at the cross-validation plot to determine K (population) size.
```{r}
#Cross validation plot
ktab<- read.table("CVerror_AC.txt", header=F)

tab<-describeBy(ktab$V2, ktab$V1, mat=T)

ggplot(tab, aes(x=as.numeric(group1), y=mean)) +
  geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=0.1) +
  geom_point(size=2)+ xlab("K")+ylab("Prediction Error")+theme_set(theme_bw())+
  scale_x_continuous(breaks = seq(0, 10, by = 1))

```

## Clumpak
After running ADMIXTURE, we can automate the analysis of the replicate runs using [clumpak](). The output for K=4 will be plotted below.
```{r}
#separate by populations
amix<- read.table("admix_k4_clumpak.txt", header=T,check.names=FALSE,stringsAsFactors = FALSE, sep="\t")

#reorganize data frame for ggplot
admix.results <- melt(amix)

admix.results$Original.pop = factor(admix.results$Original.pop, levels=c("Florida","Cuba","Belize","San Andrés","Urabá Gulf","San Bernardo","Rosario","I. Arena","Tayrona","Curacao","Puerto Rico","USVI"))

p <- ggplot(admix.results, aes(x=sample.id, y=value, fill=variable))
p <- p + geom_bar(stat='identity',color="darkgrey",size=0.0001)
p <- p + facet_grid(~ Original.pop, scales = "free_x",space = "free_x")
p <- p + theme(axis.text.x = element_blank(), axis.ticks.x=element_blank(), panel.spacing = unit(0, "lines"))+
    scale_y_continuous(expand = c(0, 0))
p <- p + scale_fill_manual(values=c(Pop1="#FCBF49",Pop2="#D62828",Pop3="#EAE2B7",Pop4="#003049"))
p

```

# Genetic Differentiation

## AMOVA
Analysis of Molecular Variance (AMOVA) is a method for estimating population differentiation from molecular data taking into account the mutational distance between alleles. Unlike Fst, which quantifies genetic differentiation based on allele frequencies, AMOVA treats molecular data as vectors and estimates Euclidean distances between alleles. Furthermore, it is possible to test hypotheses about differentiation by grouping subpopulations in a hierarchical structure (Excoffier et al., 1992).
```{r}
gindAC <- vcfR2genind(ACvcf)
gindAC@pop <- as.factor(poptab$region2)
strata(gindAC) <- data.frame(poptab[,c(4,7,8,9,10)])
gindAC

table(strata(gindAC, ~region2/subregion2))

xdis<-bitwise.dist(gindAC)

amova <- poppr.amova(gindAC, ~region2,  dist=xdis)

amova

set.seed(1999)
signif <- randtest(amova, nrepet = 9999)
signif
plot(signif)

amova2 <- poppr.amova(gindAC, ~assigned_pop/region2,  dist=xdis)

amova2

set.seed(1999)
signif2 <- randtest(amova2, nrepet = 9999)
signif2
plot(signif2)
```
There is significant evidence of population structure by region (not surprising).

## FST
Fst is a measure of average pairwise distances between pairs of individuals (haplotypes) based on allele frequencies. Population comparisons of Fst are then considered the "difference" between allele frequencies of two populations. The values range from 0 to 1, where zero is when the allele frequencies are identical in the two species/populations whereas a value of one implies that all genetic variation is explained by the population structure ("fixed" difference). Practically speaking FST values are never as high as 1 because these values are only utilized to measure breeding between populations of the same species. Different populations of the same species by definition would show some evidence of interbreeding, even if that breeding may no longer occur today.

We will calculate the matrices of genetic distances among individuals and populations. The _stamppFst_ function calculates the pairwise Fst values along with confidence intervals and p-values between populations according to the method proposed by Wright(1949) and updated by Weir and Cockerham (1984). The _basic.stats_ function estimates individual counts, allelic frequencies, observed heterozygosities and genetic diversities per locus and population (Fst and Fstp as well as Fis following Nei (1987) per locus and overall loci).

```{r}
#Weir and Cockerham (1984) formula
sf<-stamppFst(gindAC, nboots = 100, percent = 95, nclusters = parallel::detectCores() - 3)

#write.table(sf,"FST_CI_values.txt", sep="\t")

#summary by region
# separate populations into individual genind objects

gindAC <- vcfR2genind(ACvcf)
gindAC@pop <- as.factor(poptab$region)
strata(gindAC) <- data.frame(poptab[,c(4,7,8,9,10)])
gindAC

n.pop <- seppop(gindAC)

bs_I<-basic.stats(n.pop$`I. Arena`)
bs_I$overall

bs_T<-basic.stats(n.pop$Tayrona)
bs_T$overall

bs_SB<-basic.stats(n.pop$`San Bernardo`)
bs_SB$overall

bs_SA<-basic.stats(n.pop$`San Andrés`)
bs_SA$overall

bs_R<-basic.stats(n.pop$Rosario)
bs_R$overall

bs_Cap<-basic.stats(n.pop$`Urabá Gulf`)
bs_Cap$overall

he_ho<-cbind(TI_He=bs_T$perloc$Hs,TI_HO=bs_T$perloc$Ho,  SB_He=bs_SB$perloc$Hs, SB_HO=bs_SB$perloc$Ho,
            R_He=bs_R$perloc$Hs,  R_HO=bs_R$perloc$Ho, SA_He=bs_SA$perloc$Hs,SA_HO=bs_SA$perloc$Ho,
            Cap_He=bs_Cap$perloc$Hs, Cap_HO=bs_Cap$perloc$Ho)

boxplot(he_ho)
```
```{r}
t.test(bs_T$perloc$Hs,bs_T$perloc$Ho, pair = T, var.equal = TRUE, alter = "greater")
```

Lump all locations together:
```{r}
# Estimates of heterozygosity for the whole species
sub_gindAC<-popsub(gindAC,
  sublist = c("Urabá Gulf","Rosario","San Andrés","San Bernardo","Tayrona", "I. Arena"),
)
toto_AC <- summary(sub_gindAC)

# plot Ho amd He Acerv estimates together
boxplot(toto_AC$Hexp, toto_AC$Hobs)
```
Test for significant difference between observed and expected heterozygosity.
```{r}
# significance test
bartlett.test(list(toto_AC$Hexp, toto_AC$Hobs))
t.test(toto_AC$Hexp, toto_AC$Hobs, pair = T, var.equal = TRUE, alter = "greater")

```

# Isolation-by-distance test
The Mantel's test assesses whether the pairwise genetic distance matrix is correlated with the pairwise geographic distance matrix.
```{r}
AC.mat2 <- read.table("geoDist2.txt", header=T,sep="\t",
                               check.names=FALSE,stringsAsFactors = FALSE)

AC.xy <- dist(cbind(AC.mat2$long, AC.mat2$lat))

#Calculate the distance between two locations/nests.
#default unit of measurement is kilometers
AC.xy  <- distm(as.matrix(AC.mat2[,c("long","lat")]), fun = distHaversine)/1000

# Euclidian genetic distance of all sites
genDist<-as.dist(sf$Fsts)

# just Colombia sites
genDist2<-as.dist(sf$Fsts[7:12,7:12])
AC.xy2<-as.dist(AC.xy[7:12,7:12])
```

We’ll create a population-level Mantel correlogram with the proportion of shared alleles. The biological hypothesis of isolation-by-distance postulates that genetic distance increases with geographic distance. In this method there is no assumption of linearity.
```{r}
AC.correlog <- mantel.correlog(genDist2, XY=AC.xy2, nperm=9999)
AC.correlog
plot(AC.correlog)

```
In this case we find the first two distance classes show positive spatial autocorrelation.

Next, run a Mantel test between the two distance matrices.
```{r}

man<-vegan::mantel(xdis = genDist2, ydis = as.dist(AC.xy2), method = "spearman") 
man

aa = as.vector(genDist2)
tt = as.vector(AC.xy2)
mat = data.frame(aa,tt)

my.formula <- y ~ x
mm = ggplot(mat, aes(y = aa, x = tt)) +
    geom_point(size = 3, alpha = 0.5) +
    geom_smooth(method = "lm", colour = "black", alpha = 0.2) +
    labs(x = "Geographic Distance (km)", y = "Genetic Distance (FST)") +
    theme( axis.text.x = element_text(face = "bold",colour = "black", size = 12),
        axis.text.y = element_text(face = "bold", size = 11, colour = "black"),
        axis.title= element_text(face = "bold", size = 14, colour = "black"),
        panel.background = element_blank(),
        panel.border = element_rect(fill = NA, colour = "black"))
mm
```

The Mantel correlation was significant, and the relationship not linear, hence the using the Spearman rank correlation rather than Pearson correlation.

### Identity-by-state

IBS is a term used in genetics to describe two identical alleles or two identical segments or sequences of DNA. IBS is generally used to describe segments which are not necessarily derived from direct descent (identical by descent (IBD)) but are shared by many individuals both within and between populations. Despite the fact that IBS markers are not clearly inherited we can still use them to determine relatedness. The initial code lines here are re-loading up the data, since we need to read it in differently for this analysis package. The identity by state analysis is at the end of the code block.

```{r}
vcf.fn <- "AC_mergeDB_maf0.05.recode.vcf"
snpgdsVCF2GDS(vcf.fn, "test.gds", method="biallelic.only")
genofile <- snpgdsOpen(filename="test.gds",  readonly=FALSE)

hd<-read.gdsn(index.gdsn(genofile, "sample.id"))
hd<-data.frame(hd)
hd<-setDT(hd, keep.rownames = FALSE)[]
setnames(hd, c("hd"), c("user_id"))

samp.annot <- data.frame(pop.group = c(poptab$region))
add.gdsn(genofile, "sample.annot", samp.annot)
samp.id <- data.frame(sample.id = c(poptab$user_specimen_id))
add.gdsn(genofile, "sample.id2", samp.id)

pop_code <- read.gdsn(index.gdsn(genofile, path="sample.annot/pop.group"))
table(pop_code)
pop.group <- as.factor(read.gdsn(index.gdsn(genofile, "sample.annot/pop.group")))
pop.level <- levels(pop.group)
samp_code <-read.gdsn(index.gdsn(genofile, path="sample.id2/sample.id"))

# individuLs in the same population are clustered together
pop.idx <- order(pop_code)

#cluster analysis on the n×nn×n matrix of genome-wide IBS pairwise distance
set.seed(100)
par(cex=0.6, cex.lab=1, cex.axis=1.5,cex.main=2)
ibs <- snpgdsIBS(genofile, num.thread=2,autosome.only=FALSE)
ibs$sample.id <-poptab$user_specimen_id
ibs.hc <- snpgdsHCluster(ibs)
```
Using the output of the identity by state analysis, we can construct a tree of the different samples. Note that this may look similar to the tree we built via the neighbor-joining approach before, but the way that the distances between the different samples were obtained was fundamentally different. We will begin by coloring based on our now so familiar different species.
```{r}
#default clustering
#pdf ("IBS_default.pdf", width=10, height=7)
rv <- snpgdsCutTree(ibs.hc, col.list=cols3, pch.list=15)
snpgdsDrawTree(rv, main="Color by Cluster", leaflab="perpendicular",y.label=0.2)
legend("topleft", legend=levels(rv$samp.group), xpd=T, col=cols3[1:nlevels(rv$samp.group)], pch=15, ncol=4, cex=1.2)
#dev.off()
```
As with every other approach, we see a very clear distinction between the different species in this tree. We will now color based on the collection location and see if we see clustering in co-ancestry within a given region.
```{r}
#region
#pdf ("IBS_Region.pdf", width=10, height=7)
race <- as.factor(pop_code)
rv2 <- snpgdsCutTree(ibs.hc,samp.group=race,col.list=cols2,pch.list=15)
snpgdsDrawTree(rv2, main="Color by Region", leaflab="perpendicular",y.label=0.2)
legend("topleft", legend=levels(race), xpd=T, col=cols2[1:nlevels(race)], pch=15, ncol=4, cex=1.2)
#dev.off

#close GDS file
snpgdsClose(genofile)
```
